pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -632,9 +691,48 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 		changes.KillPod = true
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarLifecycle) {
+		if pod.Spec.RestartPolicy != v1.RestartPolicyAlways {
+			//Determine if there are any non sidecar containers that are running or need restarting
+			//if there are none, we can kill the remaining sidecars
+			onlySidecars := onlySidecars(pod, podStatus)
+
+			//only sidecars are left so terminate them all
+			if onlySidecars {
+				changes.KillPod = true
+				for idx, container := range pod.Spec.Containers {
+					containerStatus := podStatus.FindContainerStatusByName(container.Name)
+					// we don't need to terminate non sidecars or exited sidecars
+					if lifecycleSidecar(container) && containerStatus.State != kubecontainer.ContainerStateExited {
+						message := " All containers have permanently exited, sidecar container will be killed"
+						changes.ContainersToKill[containerStatus.ID] = containerToKillInfo{
+							name:      containerStatus.Name,
+							container: &pod.Spec.Containers[idx],
+							message:   message,
+						}
+						klog.V(2).Infof("Container %q (%q) of pod %s: %s", container.Name, containerStatus.ID, format.Pod(pod), message)
+					}
+					//Remove this container from the list of containers to start
+					for i, id := range changes.ContainersToStart {
```

Comment:
Should this for loop be inside the if block above which starts on line 706 ?

Reply:
No this is intentional, this is to ensure that the kubelet does not try to restart any sidecars if they are the only ones left.